landlab/components/overland_flow/generate_overland_flow_implicit_kinwave.py

```diff
@@ -0,0 +1,320 @@
+
+            # Re-route flow, which gives us the downstream-to-upstream
+            # ordering
+            self.flow_accum.run_one_step()
+            self.nodes_ordered = self.grid.at_node['flow__upstream_node_order']
+            self.flow_lnks = self.grid.at_node['flow__links_to_receiver_nodes']
+
+            # (Re)calculate, for each node, sum of sqrt(gradient) x width
+            self.grad_width_sum[:] = 0.0
+            for i in range(self.flow_lnks.shape[1]):
+                self.grad_width_sum[:] += (self.sqrt_slope[self.flow_lnks[:,i]]
+                    * self._grid.width_of_face[
+                            self.grid.face_at_link[self.flow_lnks[:,i]]])
+
+            # Calculate values of alpha, which is defined as
+            #
+            #   $\alpha = \frac{\Sigma W S^{1/2} \Delta t}{A C_r}$
+            cores = self.grid.core_nodes
+            self.alpha[cores] = (
+                self.vel_coef * self.grad_width_sum[cores] * dt 
+                / (self.grid.area_of_cell[self.grid.cell_at_node[cores]]))
+
+        # Zero out inflow discharge
+        self.disch_in[:] = 0.0
+
+        # Upstream-to-downstream loop
+        for i in range(len(self.nodes_ordered) - 1, -1, -1):
+            n = self.nodes_ordered[i]
+            if self.grid.status_at_node[n] == 0:
+
+                # Solve for new water depth
+                aa = self.alpha[n]
+                cc = self.depth[n]
+                ee = ((dt * runoff_rate) 
+                       + (dt * self.disch_in[n] 
+                          / self.grid.area_of_cell[self.grid.cell_at_node[n]]))
+                self.depth[n] = newton(water_fn, self.depth[n], 
+                                       args=(aa, self.weight, cc, 
+                                             self.depth_exp, ee))
+
+                # Calc outflow
+                Heff = (self.weight * self.depth[n]
+                        + (1.0 - self.weight) * cc)
+                outflow = (self.vel_coef * (Heff ** self.depth_exp)
+                           * self.grad_width_sum[n])  # this is manning/chezy/darcy
+
+                # Send flow downstream. Here we take total inflow discharge
+                # and partition it among the node's neighbors. For this, we use
+                # the flow director's "proportions" array, which contains, for
+                # each node, the proportion of flow that heads out toward each
+                # of its N neighbors. The proportion is zero if the neighbor is
+                # uphill; otherwise, it is S^1/2 / sum(S^1/2). If for example
+                # we have a raster grid, there will be four neighbors and four
+                # proportions, some of which may be zero and some between 0 and
+                # 1.
+                self.disch_in[self.grid.neighbors_at_node[n]] += (outflow
+                    * self.flow_accum.flow_director.proportions[n])
+
+                # TODO: the above is enough to implement the solution for flow
```

Comment:
I agree that we want to eventually included flow velocity and discharge on links.